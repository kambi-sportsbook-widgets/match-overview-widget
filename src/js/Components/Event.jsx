import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { translationModule, widgetModule } from 'kambi-widget-core-library'
import { OutcomeButton, OutcomeButtonUI } from 'kambi-widget-components'
import styles from './Event.scss'

/**
 * Translation helper
 * @type {function}
 */
const t = translationModule.getTranslation.bind(translationModule)

/**
 * Pads with leading 0s to ensure number is two digit.
 * @param {number} n Number to pad
 */
const pad = function(n) {
  return n < 10 ? '0' + n : n
}

/**
 * Capitalizes first letter of a string.
 * @param {string} s Given string
 */
const cap = function(s) {
  return s.charAt(0).toUpperCase() + s.slice(1)
}

class Event extends Component {
  /**
   * Constructs.
   * @param {object} props Component properties
   */
  constructor(props) {
    super(props)
    this.onClick = this.onClick.bind(this)
  }

  /**
   * Handles click on event box.
   */
  onClick() {
    if (this.props.event.tags.indexOf('OPEN_FOR_LIVE') !== -1) {
      widgetModule.navigateToLiveEvent(this.props.event.id)
    } else {
      widgetModule.navigateToEvent(this.props.event.id)
    }
  }

  /**
   * Formatted start date.
   * @returns {string}
   */
  get startDate() {
    const now = new Date(),
      date = new Date(this.props.event.start)

    const soonStr = (function() {
      if (now.getDate() === date.getDate()) {
        return t('today')
      } else if (now.getDate() === date.getDate() - 1) {
        return t('tomorrow')
      } else {
        return ''
      }
    })()

    return (
      soonStr +
      ' ' +
      pad(date.getDate()) +
      ' ' +
      // @todo: add translations for months
      cap(t('month' + date.getMonth()).slice(0, 3)) +
      ' ' +
      pad(date.getHours()) +
      ':' +
      pad(date.getMinutes())
    )
  }

  /**
   * Renders component.
   * @returns {XML}
   */
  render() {
    // AMERICAN_DISPLAY_FORMAT means the match should show awayName first
    // it is usually only used for american football
    const americanDisplayFormat =
      this.props.event.tags.indexOf('AMERICAN_DISPLAY_FORMAT') !== -1
    const homeName = (
      <div
        className={[
          styles.team,
          americanDisplayFormat ? styles.right : styles.left,
        ].join(' ')}
      >
        {this.props.event.homeFlag && (
          <img
            className={styles.flag}
            src={this.props.event.homeFlag.url}
            alt=""
          />
        )}
        <span className={styles.name}>{this.props.event.homeName}</span>
      </div>
    )

    const awayName = (
      <div
        className={[
          styles.team,
          americanDisplayFormat ? styles.left : styles.right,
        ].join(' ')}
      >
        <span className={styles.name}>{this.props.event.awayName}</span>
        {this.props.event.awayFlag && (
          <img
            className={styles.flag}
            src={this.props.event.awayFlag.url}
            alt=""
          />
        )}
      </div>
    )

    let outcomes = this.props.outcomes
    if (americanDisplayFormat) {
      if (outcomes.length === 3) {
        outcomes = [outcomes[2], outcomes[1], outcomes[0]]
      } else if (outcomes.length === 2) {
        outcomes = [outcomes[1], outcomes[0]]
      }
    }

    const score = this.props.liveData ? this.props.liveData.score : null
    return (
      <div className={styles.general}>
        <div className={styles.header} onClick={this.onClick}>
          <div className={styles.group}>{this.props.event.group}</div>
          <div className={styles.start}>{this.startDate}&nbsp;</div>
        </div>
        <div className={styles.teams} onClick={this.onClick}>
          {americanDisplayFormat ? awayName : homeName}

          {this.props.liveData && (
            <div className={styles.score}>
              {americanDisplayFormat ? score.away : score.home}
              <small>-</small>
              {americanDisplayFormat ? score.home : score.away}
            </div>
          )}

          {americanDisplayFormat ? homeName : awayName}
        </div>
        <div className={`${styles.outcomes}`}>
          {!this.props.liveData &&
            outcomes.map(outcome => (
              <OutcomeButton
                key={outcome.id}
                outcome={outcome}
                event={this.props.event}
                updateOdds={false}
              />
            ))}

          {this.props.liveData && (
            <OutcomeButtonUI
              label={
                <span className={styles.liveLabel}>
                  <em>{t('Live')}</em>
                  {t('Right Now')}
                </span>
              }
              onClick={this.onClick}
            />
          )}
        </div>
      </div>
    )
  }
}

Event.propTypes = {
  /**
   * Event object
   */
  event: PropTypes.object.isRequired,

  /**
   * Array of outcomes
   */
  outcomes: PropTypes.array,

  /**
   * Live data object
   */
  liveData: PropTypes.object,
}

Event.defaultProps = {
  outcomes: [],
  liveData: null,
}

export default Event
