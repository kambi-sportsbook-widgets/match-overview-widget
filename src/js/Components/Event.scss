*:last-child > :local(.general) {
   border-right: 0;
}

:local(.general) {
   border-right: 1px dotted rgba(255, 255, 255, 0.4);
   margin-top: 6px;
   max-width: 350px;
   width: 350px;
   height: 120px;
}

:local(.header) {
   display: flex;
   padding-top: 6px;
   opacity: .9;
   font-size: 12px;
   line-height: 24px;

   :local(.group) {
      flex: 1;
      padding-left: 16px;
      justify-content: flex-start;
      align-items: center;
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
   }

   :local(.start) {
      padding-right: 16px;
      justify-content: flex-end;
      align-items: center;
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
   }
}

:local(.teams) {
   padding: 6px 16px 12px 16px;
   flex: 1;
   display: flex;
}

:local(.team) {
   flex: 1 1 auto;
   display: flex;
   align-items: center;

   .flag {
      margin-right: 12px;
      width: 24px;
      height: 24px;
   }

   .name {
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
      font-size: 14px;
      line-height: 24px;
   }

   &.home {
      justify-content: flex-start;
      margin-right: 6px;
   }

   &.away {
      justify-content: flex-end;
      margin-left: 6px;
   }
}

:local(.outcomes) {
   display: flex;
   margin-left: 16px;
   margin-right: 16px;

   > * {
      margin-left: 6px;
      flex: 1;

      &:first-child {
         margin-left: 0;
      }
   }
}
