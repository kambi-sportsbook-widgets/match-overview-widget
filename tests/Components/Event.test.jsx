/* eslint-env jest */
import React, { Children } from 'react';
import ReactTestUtils from 'react-addons-test-utils';
import { shallow } from 'enzyme';
import Event from '../../src/js/Components/Event';
import { widgetModule } from 'kambi-widget-core-library';

jest.mock('kambi-widget-core-library', () => ({
   widgetModule: {
      navigateToEvent: jest.fn(),
      navigateToLiveEvent: jest.fn()
   },
   translationModule: {
      getTranslation: jest.fn((key) => `Translated ${key}`)
   }
}));

let renderer;

const DAY = 24 * 60 * 60 * 1000;

const mockEvent = {
   id: 100,
   start: 10 * DAY,
   group: 'Test group',
   homeName: 'Sweden',
   homeFlag: {
      url: 'homeFlagUrl'
   },
   awayName: 'Poland',
   awayFlag: {
      url: 'awayFlagUrl'
   },
   openForLiveBetting: false
};

const mockLiveData = {
   score: {
      home: 1,
      away: 5
   }
};

const mockOutcomes = [
   {id: 200},
   {id: 300}
];

const originalDate = window.Date;

describe('Event DOM rendering', () => {

   beforeEach(() => {
      renderer = ReactTestUtils.createRenderer();

      window.Date = function(input = 10 * DAY) {

         this.date = new originalDate(input);

         this.getDate = () => this.date.getDate();
         this.getMonth = () => this.date.getMonth();
         this.getHours = () => this.date.getHours();
         this.getMinutes = () => this.date.getMinutes();
      };

      window.Date.now = () => {
         return 10 * DAY;
      };

   });

   afterEach(() => {
      window.Date = originalDate;
   });

   it('renders correctly with default props', () => {
      expect(renderer.render(
         <Event event={mockEvent} />
      )).toMatchSnapshot();
   });

   it('renders correctly with event starting today', () => {
      expect(renderer.render(
         <Event
            event={Object.assign(mockEvent, {start: 10 * DAY})}
         />
      )).toMatchSnapshot();
   });

   it('renders correctly with event starting tomorrow', () => {
      expect(renderer.render(
         <Event
            event={Object.assign(mockEvent, {start: 11 * DAY})}
         />
      )).toMatchSnapshot();
   });

   it('renders correctly with event starting later than tomorrow', () => {
      expect(renderer.render(
         <Event
            event={Object.assign(mockEvent, {start: 12 * DAY})}
         />
      )).toMatchSnapshot();
   });

   it('renders correctly with live data', () => {
      expect(renderer.render(
         <Event
            event={mockEvent}
            liveData={mockLiveData}
         />
      )).toMatchSnapshot();
   });

   it('renders correctly with outcomes', () => {
      expect(renderer.render(
         <Event
            event={mockEvent}
            outcomes={mockOutcomes}
         />
      )).toMatchSnapshot();
   });

});

describe('Event behaviour', () => {

   it('handles clicks correctly when not in live mode', () => {
      const wrapper = shallow(
         <Event event={mockEvent} />
      );

      expect(widgetModule.navigateToEvent).not.toHaveBeenCalled();

      wrapper.find('.header').simulate('click');

      expect(widgetModule.navigateToEvent).toHaveBeenCalledTimes(1);
      expect(widgetModule.navigateToEvent).toHaveBeenLastCalledWith(mockEvent.id);
   });

   it('handles clicks correctly when in live mode', () => {
      const wrapper = shallow(
         <Event event={Object.assign(mockEvent, {openForLiveBetting: true})} />
      );

      expect(widgetModule.navigateToLiveEvent).not.toHaveBeenCalled();

      wrapper.find('.header').simulate('click');

      expect(widgetModule.navigateToLiveEvent).toHaveBeenCalledTimes(1);
      expect(widgetModule.navigateToLiveEvent).toHaveBeenLastCalledWith(mockEvent.id);
   });

});
