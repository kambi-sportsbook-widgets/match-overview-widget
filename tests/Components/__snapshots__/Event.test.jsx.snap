// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event DOM rendering renders correctly with default props 1`] = `
<div
  className="general"
>
  <div
    className="header"
    onClick={[Function]}
  >
    <div
      className="group"
    >
      Test group
    </div>
    <div
      className="start"
    >
      Translated today 09 Tra 09:00
       
    </div>
  </div>
  <div
    className="teams"
    onClick={[Function]}
  >
    <div
      className="team home"
    >
      <img
        alt=""
        className="flag"
        src="homeFlagUrl"
      />
      <span
        className="name"
      >
        Sweden
      </span>
    </div>
    <div
      className="team away"
    >
      <span
        className="name"
      >
        Poland
      </span>
      <img
        alt=""
        className="flag"
        src="awayFlagUrl"
      />
    </div>
  </div>
  <div
    className="outcomes"
  />
</div>
`;

exports[`Event DOM rendering renders correctly with event starting later than tomorrow 1`] = `
<div
  className="general"
>
  <div
    className="header"
    onClick={[Function]}
  >
    <div
      className="group"
    >
      Test group
    </div>
    <div
      className="start"
    >
       11 Tra 09:00
       
    </div>
  </div>
  <div
    className="teams"
    onClick={[Function]}
  >
    <div
      className="team home"
    >
      <img
        alt=""
        className="flag"
        src="homeFlagUrl"
      />
      <span
        className="name"
      >
        Sweden
      </span>
    </div>
    <div
      className="team away"
    >
      <span
        className="name"
      >
        Poland
      </span>
      <img
        alt=""
        className="flag"
        src="awayFlagUrl"
      />
    </div>
  </div>
  <div
    className="outcomes"
  />
</div>
`;

exports[`Event DOM rendering renders correctly with event starting today 1`] = `
<div
  className="general"
>
  <div
    className="header"
    onClick={[Function]}
  >
    <div
      className="group"
    >
      Test group
    </div>
    <div
      className="start"
    >
      Translated today 09 Tra 09:00
       
    </div>
  </div>
  <div
    className="teams"
    onClick={[Function]}
  >
    <div
      className="team home"
    >
      <img
        alt=""
        className="flag"
        src="homeFlagUrl"
      />
      <span
        className="name"
      >
        Sweden
      </span>
    </div>
    <div
      className="team away"
    >
      <span
        className="name"
      >
        Poland
      </span>
      <img
        alt=""
        className="flag"
        src="awayFlagUrl"
      />
    </div>
  </div>
  <div
    className="outcomes"
  />
</div>
`;

exports[`Event DOM rendering renders correctly with event starting tomorrow 1`] = `
<div
  className="general"
>
  <div
    className="header"
    onClick={[Function]}
  >
    <div
      className="group"
    >
      Test group
    </div>
    <div
      className="start"
    >
      Translated tomorrow 10 Tra 09:00
       
    </div>
  </div>
  <div
    className="teams"
    onClick={[Function]}
  >
    <div
      className="team home"
    >
      <img
        alt=""
        className="flag"
        src="homeFlagUrl"
      />
      <span
        className="name"
      >
        Sweden
      </span>
    </div>
    <div
      className="team away"
    >
      <span
        className="name"
      >
        Poland
      </span>
      <img
        alt=""
        className="flag"
        src="awayFlagUrl"
      />
    </div>
  </div>
  <div
    className="outcomes"
  />
</div>
`;

exports[`Event DOM rendering renders correctly with live data 1`] = `
<div
  className="general"
>
  <div
    className="header"
    onClick={[Function]}
  >
    <div
      className="group"
    >
      Test group
    </div>
    <div
      className="start"
    >
       11 Tra 09:00
       
    </div>
  </div>
  <div
    className="teams"
    onClick={[Function]}
  >
    <div
      className="team home"
    >
      <img
        alt=""
        className="flag"
        src="homeFlagUrl"
      />
      <span
        className="name"
      >
        Sweden
      </span>
    </div>
    <div
      className="score"
    >
      1
      <small>
        -
      </small>
      5
    </div>
    <div
      className="team away"
    >
      <span
        className="name"
      >
        Poland
      </span>
      <img
        alt=""
        className="flag"
        src="awayFlagUrl"
      />
    </div>
  </div>
  <div
    className="outcomes"
  >
    <OutcomeButtonUI
      label={
        <span
          className="liveLabel"
        >
          <em>
            Translated Live
          </em>
          Translated Right Now
        </span>
      }
      odds={null}
      onClick={[Function]}
      selected={false}
      suspended={false}
    />
  </div>
</div>
`;

exports[`Event DOM rendering renders correctly with outcomes 1`] = `
<div
  className="general"
>
  <div
    className="header"
    onClick={[Function]}
  >
    <div
      className="group"
    >
      Test group
    </div>
    <div
      className="start"
    >
       11 Tra 09:00
       
    </div>
  </div>
  <div
    className="teams"
    onClick={[Function]}
  >
    <div
      className="team home"
    >
      <img
        alt=""
        className="flag"
        src="homeFlagUrl"
      />
      <span
        className="name"
      >
        Sweden
      </span>
    </div>
    <div
      className="team away"
    >
      <span
        className="name"
      >
        Poland
      </span>
      <img
        alt=""
        className="flag"
        src="awayFlagUrl"
      />
    </div>
  </div>
  <div
    className="outcomes"
  >
    <OutcomeButton
      event={
        Object {
          "id": 100,
        }
      }
      label={true}
      outcome={
        Object {
          "id": 200,
        }
      }
      outlineStyle={false}
      updateOdds={false}
    />
    <OutcomeButton
      event={
        Object {
          "id": 100,
        }
      }
      label={true}
      outcome={
        Object {
          "id": 300,
        }
      }
      outlineStyle={false}
      updateOdds={false}
    />
  </div>
</div>
`;
